### PR TITLE
MAINT: Simplify element setting and use it for filling

### PR DESCRIPTION
**Pull request by seberg**
This slightly modifies the behaviour of `arr.fill()` to be
`arr.fill(scalar)`, i.e. match `arr1d[0] = scalar`, rather than
`arr.fill(np.asarray(scalar))`, which subtely different!
(Note that object was already special cased to have the scalar
logic.)

Otherwise, `PyArray_Pack` is now the actual, full featured, "scalar"
assignment logic.  It is a bit strange due to that quantity/masked
array issue, but there is nothing to be done about it.
The simplifications in `PyArray_AssignFromCache` should not cause
any change in practice, because non 0-D arrays would have been
rejected earlier on in that path.
(Basically, it does not need the full `PyArray_Pack` logic, but that
is fine, I intially split the two, but consolidated them again.)

### CODE REVIEW

**Comment by seberg** [comment]
This is also fairly important, since it finally full consolidates what it means to set an item (at least to the extend possible).  We also really should add `PyArray_Pack` to the public API in order to nudge people away from `PyArray_SETITEM` (although, effectively, I think we can just call it from the setitem function for the user...).
(There is a bit of a problem with `HPy` probably wanting an `owner` in case the descriptor has references, we could just add an `owner` to the signature, but it feels a bit weird if we don't even know how to use it correctly everywhere yet.)

This means that a followup will be to change *all* `PyArray_SETITEM` calls, though.

**Comment by mattip** [comment]
Didn't we discuss this at the triage meeting and decide to go ahead with it?

**Comment by seberg** [comment]
Yeah, although there is that weird thing about whether or not to unpack `np.empty((), dtype=object)` (0-D object array).  This handles the unpacking "correctly" now.

I think we should just do this and go ahead, but I do need to ping the list and we may want to follow-up on the precise route.

**Comment by mattip** [comment]
Did the answers on the list help?

**Comment by seberg** [comment]
Not quite settled, considering that Stéfan and you seem to have different opinions :).  I would go with this (i.e. store the array directly), for now.  Not because I am 100% sure that this is the best way, but because my guess is that it is so much more common for us to do that (item assignment), that it probably barely matters to go ahead and change `fill` even if we want to flip it all again.
And I still think it is more important to try to have a "one right way" when it comes to item assignment.

**Comment by seberg** [comment]
I would like to go for this now that we are early in the next release cycle.  Yes, there is some uncertainty about object arrays, but I prefer the consistency for now, and worry about flipping the switch later (rather than caring about keeping this one corner case on potential future behavoir).

File: doc/release/upcoming_changes/20924.compatibility.rst [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+``np.array([0]).fill(1e300)`` or ``np.array([0]).fill(np.nan)``.
+Previously, ``arr.fill()`` would unpack 0-D object arrays if the input
+array was also of object dtype, the 0-D array will now be stored as is::
+
+    >>> arr = np.array([0], dtype=object)
+    >>> arr.fill(np.array(None, dtype=object))
+    >>> arr
+    array([array(None, dtype=object)], dtype=object)
+
+Rather than ``array([None], dtype=object)``.
```
Comment (mattip) [inline doc/release/upcoming_changes/20924.compatibility.rst:23]: Maybe just
```suggestion
Previously casting may have produced slightly different answers when using values that could not be represented in the target `dtype` or when using `object`s.
```

**Comment by mattip** [comment]
[Here](https://mail.python.org/archives/list/numpy-discussion@python.org/thread/HDFBMBMM6LBQKK2IU3ZAL7F3CU3XWLLT/) is the mailing list discussion. Can we add some documentation to `np.fill` to point out how using an `object` array will return a different result than when using assignment?

File: numpy/core/_add_newdocs.py [hunk 2]
```diff
@@ -3437,6 +3437,24 @@
...
+    >>> a.fill(np.array(3))
+    >>> a
+    array([array(3), array(3)], dtype=object)
+
+    Where other forms of assignments will unpack the array being assigned:
+
+    >>> a[...] = np.array(3)
+    >>> a
+    array([3, 3], dtype=object)
+
```
Comment (seberg) [inline numpy/core/_add_newdocs.py:3457]: Hmmm, added this chunk as an example (and tweaked the release not a tiny bit).  Not sure whether it isn't a bit much, but I guess it is just at the end of the examples.

Comment (mattip) [reply numpy/core/_add_newdocs.py:3457]: +1, thanks

**Comment by mattip** [comment]
Thanks @seberg

**Pull request merged**